### PR TITLE
Add convenience methods for redirecting to another Razor Page

### LINKED ISSUE #5956: Add convenience methods for redirecting to another Razor Page without having to pass the full URL
e.g. `RedirectToPage(pagePath: "Customers/Edit.cshtml", routeArgs: new { id = 1 })`

Should be considered as part of #5921

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #5956

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/Resources.Designer.cs [hunk 1]
```diff
@@ -15,304 +15,266 @@ internal static class Resources
         /// </summary>
         internal static string ArgumentCannotBeNullOrEmpty
         {
-            get { return GetString("ArgumentCannotBeNullOrEmpty"); }
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/Resources.Designer.cs]: Were these changes made by you or when they were regenerated?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcCoreBuilderExtensions.cs [hunk 2]
```diff
@@ -53,15 +53,25 @@ public static IMvcCoreBuilder AddRazorPages(this IMvcCoreBuilder builder)
...
+            // Action Invoker
+            //
             services.TryAddEnumerable(
                 ServiceDescriptor.Singleton<IActionDescriptorProvider, PageActionDescriptorProvider>());
+            services.TryAddSingleton<IActionDescriptorChangeProvider, PageActionDescriptorChangeProvider>();
 
             services.TryAddEnumerable(
                 ServiceDescriptor.Singleton<IActionInvokerProvider, PageActionInvokerProvider>());
 
+            //
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcCoreBuilderExtensions.cs]: Does this commenting style match how we annotate things in other AddServices methods? If so we should make the comments below match, if not we should make these parts normal comments.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs [hunk 3]
```diff
@@ -41,6 +42,10 @@ static PageLoggerExtensions()
                3,
                "Request was short circuited at page filter '{PageFilter}'.");
 
+            _redirectToPageResultExecuting = LoggerMessage.Define<string>(
+                LogLevel.Information,
+                1,
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs]: @BrennanConroy says these should be unique within the package. If you make the ones you're adding/already editing unique I'll file a bug to fix the others.

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/Resources.Designer.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/Resources.Designer.cs]: The generator changed 😢

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcCoreBuilderExtensions.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcCoreBuilderExtensions.cs]: Yeah, I copied the style \ organization format from MvcCoreServices

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/Resources.Designer.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/Resources.Designer.cs]: Just do this in a separate PR and #YOLO it. You know you want those sweet sweet #PranavPoints

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcCoreBuilderExtensions.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcCoreBuilderExtensions.cs]: I'm **super** ambivalent about this.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs]: @pranavkm - please follow our established convention. If there's an issue with our IDs then we will deal with it holistically when/if that bug is triaged.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs [hunk 4]
```diff
@@ -111,7 +119,7 @@ public PageLogScope(ActionDescriptor action)
 
             public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
             {
-                for (int i = 0; i < Count; ++i)
+                for (var i = 0; i < Count; ++i)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs]: dope

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs (hunk 3 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs]: I messed up the copy-pasta. @ryanbrandenburg filed a bug to make sure we aren't messing up elsewhere.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Page.cs [hunk 5]
```diff
@@ -139,6 +139,78 @@ protected RedirectResult Redirect(string url)
...
+            => RedirectToPagePermanent(page, routeValues: null, fragment: fragment);
+
+        /// <summary>
+        /// Redirects (<see cref="StatusCodes.Status301MovedPermanently"/>) to the specified <paramref name="page"/>.
+        /// </summary>
+        /// <param name="page">The name of the page.</param>
+        /// <param name="routeValues">The parameters for a route.</param>
+        /// <param name="fragment">The fragment to add to the URL.</param>
+        /// <returns>The <see cref="RedirectToPageResult"/> with <see cref="RedirectToPageResult.Permanent"/> set.</returns>
+        protected RedirectToPageResult RedirectToPagePermanent(string page, object routeValues, string fragment)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Page.cs]: Do these mirror the set of methods we have on controller for `RedirectToAction`?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageModel.cs [hunk 6]
```diff
@@ -177,6 +177,78 @@ protected internal RedirectResult Redirect(string url)
...
+
+        /// <summary>
+        /// Redirects (<see cref="StatusCodes.Status301MovedPermanently"/>) to the specified <paramref name="page"/>.
+        /// </summary>
+        /// <param name="page">The name of the page.</param>
+        /// <param name="routeValues">The parameters for a route.</param>
+        /// <param name="fragment">The fragment to add to the URL.</param>
+        /// <returns>The <see cref="RedirectToPageResult"/> with <see cref="RedirectToPageResult.Permanent"/> set.</returns>
+        protected internal RedirectToPageResult RedirectToPagePermanent(string page, object routeValues, string fragment)
+            => new RedirectToPageResult(page, routeValues, permanent: true, fragment: fragment);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageModel.cs]: This is gross without `{ .. }`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    /// Targets a registered route.
+    /// </summary>
+    public class RedirectToPageResult : ActionResult, IKeepTempDataResult
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RedirectToPageResult"/> with the values
+        /// provided.
+        /// </summary>
+        /// <param name="pageName">The page to redirect to.</param>
+        public RedirectToPageResult(string pageName)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs]: Should this be `page` or `pageName`? it would be nice to be consistent across usages.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+        /// <summary>
+        /// Gets or sets the fragment to add to the URL.
+        /// </summary>
+        public string Fragment { get; set; }
+
+        /// <summary>
+        /// Gets or sets the protocol for the URL, such as "http" or "https".
+        /// </summary>
+        public string Protocol { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs]: Is this used anywhere?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs]: If we're going to put protocol here should we also have `host`.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (string.IsNullOrEmpty(PageName))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(PageName));
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs]: This should be `InvalidOperationException`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Page.cs (hunk 5 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Page.cs]: Yup. Although I realized the doc comments on it are a bit nicer. I'll update

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageModel.cs (hunk 6 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageModel.cs]: 🙉

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Resources.resx [hunk 10]
```diff
@@ -141,4 +141,7 @@
   <data name="PathMustBeAnAppRelativePath" xml:space="preserve">
     <value>Path must be an application relative path that starts with a forward slash '/'.</value>
   </data>
+  <data name="NoRoutesMatched" xml:space="preserve">
+    <value>No route matches the supplied values.</value>
+  </data>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Resources.resx]: Should this say `no page matches`? Should this include what the route values were?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/UrlHelperExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+using System;
+using Microsoft.AspNetCore.Routing;
+
+namespace Microsoft.AspNetCore.Mvc
+{
+    /// <summary>
+    /// Razor Page specific extensions for <see cref="IUrlHelper"/>.
+    /// </summary>
+    public static class UrlHelperExtensions
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/UrlHelperExtensions.cs]: Beautiful

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs (hunk 7 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs]: `pageName` since that's what we'd have in route values once https://github.com/aspnet/Mvc/pull/6053 goes in.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/UrlHelperExtensions.cs (hunk 11 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/UrlHelperExtensions.cs]: This should have a different name, so that it doesn't overlap with the class in Mvc.Core

File: src/Microsoft.AspNetCore.Mvc.RazorPages/UrlHelperExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            if (urlHelper == null)
+            {
+                throw new ArgumentNullException(nameof(urlHelper));
+            }
+
+            var routeValues = new RouteValueDictionary(values)
+            {
+                ["page"] = page
+            };
+            return urlHelper.RouteUrl(
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/UrlHelperExtensions.cs]: There's some more behavior that's needed here. Basically we need to preserve the value of `page` from ambient values if one wasn't specified. See https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs#L78

File: src/Microsoft.AspNetCore.Mvc.RazorPages/UrlHelperExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    /// </summary>
+    public static class UrlHelperExtensions
+    {
+        /// <summary>
+        /// Generates a URL with an absolute path for the specified <paramref name="page"/>.
+        /// </summary>
+        /// <param name="urlHelper">The <see cref="IUrlHelper"/>.</param>
+        /// <param name="page">The page to generate the url for.</param>
+        /// <returns>The generated URL.</returns>
+        public static string Page(this IUrlHelper urlHelper, string page)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/UrlHelperExtensions.cs]: We also need one of this https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs#L16

It gets the URL of the current page.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/UrlHelperExtensionsTest.cs [hunk 14]
```diff
@@ -0,0 +1,190 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Routing;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Test
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/UrlHelperExtensionsTest.cs]: Drop `.Test`

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/Resources.Designer.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Properties/Resources.Designer.cs]: Sure. I'll push it to dev and rebase.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageUrlHelperExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    /// </summary>
+    public static class PageUrlHelperExtensions
+    {
+        /// <summary>
+        /// Generates a URL with an absolute path for the specified <paramref name="page"/>.
+        /// </summary>
+        /// <param name="urlHelper">The <see cref="IUrlHelper"/>.</param>
+        /// <param name="page">The page to generate the url for.</param>
+        /// <returns>The generated URL.</returns>
+        public static string Page(this IUrlHelper urlHelper, string page)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageUrlHelperExtensions.cs]: `pageName`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        /// </summary>
+        public string Fragment { get; set; }
+
+        /// <summary>
+        /// Gets or sets the protocol for the URL, such as "http" or "https".
+        /// </summary>
+        public string Protocol { get; set; }
+
+        /// <summary>
+        /// Gets os sets the host name of the URL.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/RedirectToPageResult.cs]: `or`

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/PageUrlHelperExtensionsTest.cs [hunk 17]
```diff
@@ -0,0 +1,250 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Routing;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Test
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/PageUrlHelperExtensionsTest.cs]: drop `.Test`
